File: sdk/ruby-holons/test/composite_test.rb
# frozen_string_literal: true

require "fileutils"
require "minitest/autorun"
require "tmpdir"
require_relative "../lib/holons/composite"

class CompositeTest < Minitest::Test
  def test_member_resolves_executable_relative_to_launcher
    Dir.mktmpdir("ruby-composite-") do |dir|
      launcher = File.join(dir, "bin", "darwin_arm64", "parent")
      member_dir = File.join(File.dirname(launcher), "holons", "ruby-node")
      member = File.join(member_dir, "observability-cascade-ruby-node")
      FileUtils.mkdir_p(member_dir)
      File.write(launcher, "#!/bin/sh\n")
      File.write(member, "#!/bin/sh\n")
      File.chmod(0o755, launcher)
      File.chmod(0o755, member)

      assert_equal member, Holons::Composite.member_from_executable(launcher, "ruby-node")
    end
  end

  def test_member_errors_when_missing
    Dir.mktmpdir("ruby-composite-") do |dir|
      launcher = File.join(dir, "bin", "darwin_arm64", "parent")
      FileUtils.mkdir_p(File.dirname(launcher))
      File.write(launcher, "#!/bin/sh\n")

      assert_raises(Errno::ENOENT) do
        Holons::Composite.member_from_executable(launcher, "ruby-node")
      end
    end
  end
end

File: sdk/js-holons/tests/composite.test.js
'use strict';

const { describe, it } = require('node:test');
const assert = require('node:assert/strict');
const fs = require('node:fs');
const os = require('node:os');
const path = require('node:path');

const composite = require('../src/composite');

describe('composite member resolution', () => {
    it('resolves an executable relative to the launcher', () => {
        const dir = fs.mkdtempSync(path.join(os.tmpdir(), 'js-composite-'));
        try {
            const launcher = path.join(dir, 'bin', 'darwin_arm64', 'parent');
            const memberDir = path.join(path.dirname(launcher), 'holons', 'node-node');
            const member = path.join(memberDir, 'observability-cascade-node-node');
            fs.mkdirSync(memberDir, { recursive: true });
            fs.writeFileSync(launcher, '#!/bin/sh\n');
            fs.writeFileSync(member, '#!/bin/sh\n');
            fs.chmodSync(launcher, 0o755);
            fs.chmodSync(member, 0o755);

            assert.equal(composite.memberFromExecutable(launcher, 'node-node'), member);
        } finally {
            fs.rmSync(dir, { recursive: true, force: true });
        }
    });

    it('errors when no executable exists', () => {
        const dir = fs.mkdtempSync(path.join(os.tmpdir(), 'js-composite-'));
        try {
            const launcher = path.join(dir, 'bin', 'darwin_arm64', 'parent');
            fs.mkdirSync(path.join(path.dirname(launcher), 'holons', 'node-node'), { recursive: true });
            fs.writeFileSync(launcher, '#!/bin/sh\n');

            assert.throws(() => composite.memberFromExecutable(launcher, 'node-node'), /no executable found/);
        } finally {
            fs.rmSync(dir, { recursive: true, force: true });
        }
    });
});
